### PR TITLE
docs: automatic generate PDK docs by LDoc

### PR DESCRIPTION
**Pull request by tzssangglass**
Signed-off-by: tzssangglass <tzssangglass@gmail.com>

Automatically generate plugin development documentation by adding comments to the lua code.

### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

TODO list

~~1. specify the generated directory~~
~~2. Support more comments, such as params, etc.~~
~~3. run in CI~~
4. add more comments in the `core` folder

Memo: run command

```
 ldoc -q -c autodocs/config.ld $PWD/apisix/core/id.lua
```

### CODE REVIEW

**Comment by tzssangglass** [comment]
I found the following two problems, if you need to trigger the CI by way of label, you need two steps

1. label the PR
2. the next commit of the PR will trigger the workflow associated with the label

Simply labeling is not enough to trigger a CI.


Next we have to consider how to push the [generated md](https://github.com/apache/apisix/actions/runs/1859959963) documents to the official website.

Right now we don't seem to support this automation and still need to do it manually.

**Review by tokers — APPROVED** [review]

File: ci/autodocs.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    apt-get -y install lua5.1 liblua5.1-0-dev
+    curl https://raw.githubusercontent.com/apache/apisix/master/utils/linux-install-luarocks.sh -sL | bash -
+    luarocks install ldoc
+}
+
+generate_docs() {
+    rm -rf autodocs/output || true
+    mkdir autodocs/output || true
+    cd autodocs/output
+    ldoc -c ../config.ld ../../apisix/core/request.lua
```
Comment (spacewander) [inline ci/autodocs.sh:32]: Better to avoid listing each file one by one.

**Comment by spacewander** [comment]
Actually, the website will fetch the doc and generate an output that is suitable for the website. We can provide a script in APISIX, and let the website code runs it.

Let's discuss the integration with the website and the doc style with @guoqqqi.

**Comment by guoqqqi** [comment]
> Let's discuss the integration with the website and the doc style with @guoqqqi.

Hi, @spacewander I have some questions here, does `the doc style` mean the formatting of the document? For example, title, blocks of code, or something like that?

**Comment by guoqqqi** [comment]
Hi, I noticed that the auto-generated files are missing licenses, do we support auto-generated directories, tables?

**Comment by tzssangglass** [comment]
> Hi, I noticed that the auto-generated files are missing licenses, do we support auto-generated directories, tables?

I remember that automatically generated files do not require a license. cc @spacewander 

yes, It can be generated automatically, can you give an example?

**Comment by spacewander** [comment]
I think we only need to add a license when redistributing the files.

**Comment by tzssangglass** [comment]
> I think we only need to add a license when redistributing the files.

ok, I will add license to template.

**Comment by tzssangglass** [comment]
cc @bzp2010

File: autodocs/generate.sh [hunk 2]
```diff
@@ -18,27 +18,30 @@
...
     rm -rf autodocs/output || true
     mkdir autodocs/output || true
     cd autodocs/output
-    ldoc -c ../config.ld ../../apisix/core/request.lua
-    ldoc -c ../config.ld ../../apisix/core/id.lua
+    path="../../apisix/core"
+    files=$(ls $path)
+    for filename in $files
+    do
+       ldoc -c ../config.ld $path/$filename
```
Comment (spacewander) [inline autodocs/generate.sh:38]: We can use `find ../../apisix/core -name "*.lua" -type f -exec ldoc -c ../config.ld  {} \;`

Comment (tzssangglass) [reply autodocs/generate.sh]: fixed

**Review by spacewander — APPROVED** [review]

**Comment by tzssangglass** [comment]
wait for merge, I will add at least one method comment for each module under `apisix/core`.

**Comment by tokers** [comment]
@tzssangglass There are some conflicts should be resolved.

**Comment by tzssangglass** [comment]
> @tzssangglass There are some conflicts should be resolved.

fixed

File: apisix/core/io.lua [hunk 3]
```diff
@@ -15,12 +15,26 @@
...
+-- @module core.io
+
 local open = io.open
 
 
 local _M = {}
 
-
+---
+-- Get etcd cluster and server version.
```
Comment (spacewander) [inline apisix/core/io.lua:28]: copy & paste err?

File: apisix/core/json.lua [hunk 4]
```diff
@@ -95,8 +100,19 @@ local delay_tab = setmetatable({data = "", force = false}, {
 })
 
 
--- this is a non-thread safe implementation
--- it works well with log, eg: log.info(..., json.delay_encode({...}))
+---
+-- Delayed encoding of input data, avoid unnecessary encode operations.
+-- This is a non-thread safe implementation, it works well with log.
```
Comment (spacewander) [inline apisix/core/json.lua:105]: I think we can avoid the non-thread description, as most of the API is not thread safe.

File: apisix/core/request.lua [hunk 5]
```diff
@@ -60,9 +64,29 @@ local function _validate_header_name(name)
     return name
 end
 
+---
+-- Returns all headers of the current request.
+-- This function returns up to 100 headers by default,
```
Comment (spacewander) [inline apisix/core/request.lua:69]: Let's update the info

Comment (tzssangglass) [reply apisix/core/request.lua]: update

File: apisix/core/json.lua (hunk 4 above)
Comment (tzssangglass) [reply apisix/core/json.lua]: update

File: apisix/core/io.lua (hunk 3 above)
Comment (tzssangglass) [reply apisix/core/io.lua]: update

**Review by spacewander — APPROVED** [review]

**Review by bzp2010 — APPROVED** [review]

**Comment by juzhiyuan** [comment]
Hi @tzssangglass, those code comments are helpful to developers! Could we generate those `.md` under the `docs` folder? It will be much easier for developers to refer to. 😉  Correct me if this is not necessary.

**Comment by tzssangglass** [comment]
> Hi @tzssangglass, those code comments are helpful to developers! Could we generate those `.md` under the `docs` folder? It will be much easier for developers to refer to. 😉 Correct me if this is not necessary.

We can generate these dynamically in CI, and I expect to do this in apisix-website. cc @bzp2010

**Pull request merged**